UPGRADING.md

```diff
@@ -0,0 +1,177 @@
+
+[12]: https://hub.docker.com/r/bitnami/keycloak/tags
+
+## Upgrade nginx (minor version)
+
+The steps are the same as postgresql and keycloak above, except look for the
+nginx image at [dockerhub][13] and replace the version under "reverse-proxy:
+image:".
+
+[13]: https://hub.docker.com/r/bitnami/nginx/tags
+
+## Upgrade postgresql (major version)
+
+A major postgresql version upgrade is quite a bit more involved than changing a
+line in a file and tagging it. There are multiple ways to do it. See [the
+official postgresql docs][14].
+
+[14]: https://www.postgresql.org/docs/15/upgrading.html
+
+### The `pg_upgrade` way
+
+Official postgresql documentation mentions [pg_upgrade][15]. The `pg_upgrade`
+tool requires the binaries of the old version and old configuration directory in
+order to perform a migration of the old data format to the new data format.
+
+To provide access to the binaries of the old postgresql image in the new image,
+copy them from the image to a new directory and mount it in the new image.
+
+Be careful not to start postgres using the bitnami image until after upgrade.
+
+In a shell with the old postgres running:
+
+- `sudo docker exec -ti deploy-database-1 /bin/bash`
+
+Copy the old binaries to a specific old binaries dir (e.g. 14):
+
+- `mkdir /bitnami/postgresql/pg14bin`
+- `cp -r /opt/bitnami/postgresql/bin/* /bitnami/postgresql/pg14bin/`
+
+Stop the old postgres:
+
+- `pg_ctl stop -D /bitnami/postgresql/data`
+
+This will kick you out of the container because it is now dead.
+
+Now run a container from the old image, with env vars, e.g.
+
+- `sudo docker compose -f compose-[version].yml run -ti database /bin/bash`
+
+Move the old data from old directory to a new specific old directory:
+
+- `mv /opt/bitnami/postgresql/data /bitnami/postgresql/data.old`
+
+Copy the old postgresql.conf, pg_hba.conf, and conf.d contents to the old data
+directory and exit the container:
+
+- `cp /opt/bitnami/postgresql/conf/*.conf /bitnami/postgresql/data.old/`
+- `cp -r /bitnami/postgresql/conf.d /bitnami/postgresql/data.old/`
+- `exit`
+
+Create a new directory for the new postgres data.
+Now we need to init db but without the pdc user, only the postgres user.
+In a shell in the container running the new postgres, with volumes mounted to
+the same locations as the old, initialize the new database:
+
+- `initdb -U postgres data.new`
+
+Perform the upgrade:
+
+- `cd /bitnami/postgresql`
+- `pg_upgrade -U postgres --old-datadir /bitnami/postgresql/data.old --new-datadir /bitnami/postgresql/data.new --old-bindir /bitnami/postgresql/pg14bin --new-bindir /opt/bitnami/postgresql/bin`
+
+[15]: https://www.postgresql.org/docs/15/pgupgrade.html
+
+### The `pg_dumpall` and `pg_restore` way
+
+See [the postgresql docs][16].
+
+The steps should be to run a second postgresql container with the new version of
+postgresql and connect from it to the first, already-running postgresql
+container.
```

Comment:
The reason it is brief is I haven't tried it this way.

Reply:
I'm deleting the paragraph until I try it. My summary does not say that a new database instance has to be running, only the container with the new version of PostgreSQL. Your summary also is confusing because if `pg_dumpall` needs to connect, then the old instance needs to be running.

Reply:
The old instance needs to be running to take a backup of it, yes. The backup should be taken with the newer version of `pg_dumpall`. The new cluster does not need to be running to use `pg_dumpall`, however, and the old cluster does not need to be running to use `psql` or `pg_restore` - the backup and restore may be separated in time.